### DISCUSSION TITLE
How should I compile 32-bit programs that can run on win7?

### DISCUSSION THREAD

**Discussion by Guozht** [rustdesk/rustdesk#2224] [Q&A]
I have successfully compiled the 64 bit program in the VS2019 - WIN10 compilation environment. The program can run normally in win10, but an error is reported in WIN7 and the program input point SWDeviceCreate is not found in CFGMGR32.dll.

Then I compile 32-bit programs

Compilation environment WIN10

VS2019

llvm-win64

Then an error is reported: OneCoreUAP.lib cannot be found.

How can I successfully compile 32-bit programs that can run on WIN7?

Is there an article that guides the compilation of 32-bit programs?

**Comment by djex** [comment]
The API functions SWDeviceCreate and SwDeviceClose do not exist in Windows 7. These are present in the IddController.c file for the IDD display driver. If you do not plan on using RustDesk on a machine without a screen you can comment out enough code to get it to compile but will break the ability of using the IDD driver. This is not a fix and only a workaround. I have an older nightly source release and so this may have changed in a more recent nightly release.

I commented out any code that called SWDeviceCreate and  SwDeviceClose. You may also need to comment out calls to DiUninstallDriverW which should be in the same file.

**Reply by Guozht** [reply]
Thank you very much for your reply.

How did the release version released in github compile? Are these codes also commented out? The official website supports win7 (32-bit 64 bit) systems and above.

**Reply by Guozht** [reply]
In addition, do you know how to successfully compile 32-bit programs?

I tried the following command:

vcpkg install libvpx:x86-windows-static libyuv:x86-windows-static opus:x86-windows-static

cargo build --release --target i686-pc-windows-msvc

However, the compilation fails all the time, indicating that the 64 bit library conflicts with the 32-bit program or the work function link in the library cannot be found.

**Reply by djex** [reply]
> How did the release version released in github compile? Are these codes also commented out? The official website supports win7 (32-bit 64 bit) systems and above.

I have this same question. I am not sure how they compiled the released 1.1.9 version that is available on the website. The released source for 1.1.9 compiles into a different older? version than what is pre-compiled on the website and is missing features the compiled website release has. The pre-compiled version on the website seems closer to 1.2.0 nightly than 1.1.9 source code available on github when comparing features of each version. My guess is that they never included the IDD driver in the website release.

> In addition, do you know how to successfully compile 32-bit programs?

Unfortunately I do not. I'm not familiar enough with rust and the compiler. I attempted to compile for generic linux and after some time of failure after failure I gave up and installed a linux VM and compiled from there which worked first try. I know it can be done all from Windows however I could not get it to work. However in your case you might try installing the i686-pc-windows-msvc tool chain using rustup and then try compiling with `--target i686-pc-windows-msvc` again.

```
rustup target add i686-pc-windows-msvc
```

You can read more here: [https://rust-lang.github.io/rustup/installation/windows.html](https://rust-lang.github.io/rustup/installation/windows.html)

**Reply by evandrmb** [reply]
> How did the release version released in GitHub compile? Are these codes also commented out? The official website supports win7 (32-bit 64-bit) systems and above.
If you find it out, could you please update it here? I'm also looking for it for quite some time
